File: BehaviorTree/Branch/Decorator/Repeater.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace FluentBehaviorTree
{
    public class Repeater : Decorator
    {
        private int n;

        public Repeater(string name, int times = 0)
        {
            this.Name = name;
            n = times;
        }

        /// <summary>
        /// Repeat n times and return
        /// </summary>
        /// <returns></returns>
        protected override Status tick()
        {
            if (n == 0)
            {             
                while (true)
                {
                    this.Result = child.Tick();

                    if (this.Result == Status.ERROR) return this.Result;
                }
            }
            else
            {
                for (int i = 0; i < n; ++i)
                {
                    this.Result = child.Tick();
                    if (this.Result == Status.ERROR) return this.Result;
                }
            }
            this.Result = Status.SUCCESS;
            return this.Result;
        }
    }
}

File: BehaviorTreesLib/Repeater.cpp
#include "Repeater.h"

namespace fluentBehaviorTree
{
	Repeater::Repeater(std::string name, int times = 0)
	{
		this->setName(name);
		mN = times;
	}

	// Repeat n times and return
	EStatus Repeater::tickNode()
	{
		if (mN == 0)
		{
			while (true)
			{
				this->setResult(mChild->tick());

				if (this->getResult() == EStatus::ERROR) return this->getResult();
			}
		}
		else
		{
			for (int i = 0; i < mN; ++i)
			{
				do
				{
					this->setResult(mChild->tick());
					if (this->getResult() == EStatus::ERROR) return this->getResult();
				} while (mChild->isOpen());
			}
		}
		this->setResult(EStatus::SUCCESS);
		return this->getResult();
	}
}